### PR TITLE
[SPARK-32222][K8s][TESTS] Add K8s IT for conf propagation

### PR DESCRIPTION
**Pull request by ScrapCodes**
### What changes were proposed in this pull request?

Added integration test - which tries to configure a log4j.properties and checks if, it is the one pickup by the driver.

### Why are the changes needed?

Improved test coverage.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

By running integration tests.

### CODE REVIEW

**Comment by ScrapCodes** [comment]
I was wondering, what I am doing wrong in my setup of minkube. I have tried 3 different versions of minikube. 1.13.1, 1.14.2, 1.15.0. All of them, failed the tests with following error.

```
KubernetesSuite:
org.apache.spark.deploy.k8s.integrationtest.KubernetesSuite *** ABORTED ***
  io.fabric8.kubernetes.client.KubernetesClientException: An error has occurred.
  at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:64)
  at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:53)
  at io.fabric8.kubernetes.client.utils.HttpClientUtils.createHttpClient(HttpClientUtils.java:234)
  at io.fabric8.kubernetes.client.utils.HttpClientUtils.createHttpClient(HttpClientUtils.java:63)
  at io.fabric8.kubernetes.client.BaseClient.<init>(BaseClient.java:51)
  at io.fabric8.kubernetes.client.DefaultKubernetesClient.<init>(DefaultKubernetesClient.java:145)
  at org.apache.spark.deploy.k8s.integrationtest.backend.minikube.Minikube$.getKubernetesClient(Minikube.scala:81)
  at org.apache.spark.deploy.k8s.integrationtest.backend.minikube.MinikubeTestBackend$.initialize(MinikubeTestBackend.scala:33)
  at org.apache.spark.deploy.k8s.integrationtest.KubernetesSuite.beforeAll(KubernetesSuite.scala:130)
  at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:212)
  ...
  Cause: java.nio.file.NoSuchFileException: /home/prashant/.minikube/apiserver.crt
  at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
  at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
  at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
  at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
  at java.nio.file.Files.newByteChannel(Files.java:361)
  at java.nio.file.Files.newByteChannel(Files.java:407)
  at java.nio.file.Files.readAllBytes(Files.java:3152)
  at io.fabric8.kubernetes.client.internal.CertUtils.getInputStreamFromDataOrFile(CertUtils.java:74)
  at io.fabric8.kubernetes.client.internal.CertUtils.createKeyStore(CertUtils.java:244)
  at io.fabric8.kubernetes.client.internal.SSLUtils.keyManagers(SSLUtils.java:128)
  ...

```

**Comment by dongjoon-hyun** [comment]
:) You are hitting at the known security breaking change from Minikube side. It occurs long time ago. Please try to do the following as a workaround, @ScrapCodes . It will unblock you.
```
$ minikube ssh "sudo cat /var/lib/minikube/certs/apiserver.key" > ~/.minikube/apiserver.key
$ minikube ssh "sudo cat /var/lib/minikube/certs/apiserver.crt" > ~/.minikube/apiserver.crt
```

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala [hunk 1]
```diff
@@ -43,8 +43,8 @@ import org.apache.spark.internal.config._
 class KubernetesSuite extends SparkFunSuite
   with BeforeAndAfterAll with BeforeAndAfter with BasicTestsSuite with SecretsTestsSuite
   with PythonTestsSuite with ClientModeTestsSuite with PodTemplateSuite with PVTestsSuite
-  with DepsTestsSuite with DecommissionSuite with RTestsSuite with Logging with Eventually
-  with Matchers {
+  with DepsTestsSuite with DecommissionSuite with RTestsSuite with SparkConfPropagateSuite
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:46]: Shall we put this after `BasicTestsSuite` or `SecretsTestsSuite`?

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/SparkConfPropagateSuite.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import scala.io.{BufferedSource, Source}
+
+import io.fabric8.kubernetes.api.model._
+
+import org.apache.spark.internal.config
+
+private[spark] trait SparkConfPropagateSuite { k8sSuite: KubernetesSuite =>
+  import KubernetesSuite.{k8sTestTag, SPARK_PI_MAIN_CLASS}
+
+  test("Verify logging configuration is picked from the provided SPARK_CONF_DIR/log4j.properties",
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/SparkConfPropagateSuite.scala:31]: Ur, is the test name correct? `log-config-test-log4j.properties` seems to be used in the test.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala (hunk 1 above)
Comment (ScrapCodes) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:46]: Yes!

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/SparkConfPropagateSuite.scala (hunk 2 above)
Comment (ScrapCodes) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/SparkConfPropagateSuite.scala:31]: That is the name to keep it distinct from existing `log4j.properties`. However, when the spark conf dir is created, the default logger file name is log4j.properties.

**Comment by ScrapCodes** [comment]
> :) You are hitting at the known security breaking change from Minikube side. It occurs long time ago. Please try to do the following as a workaround, @ScrapCodes . It will unblock you.
> 
> ```
> $ minikube ssh "sudo cat /var/lib/minikube/certs/apiserver.key" > ~/.minikube/apiserver.key
> $ minikube ssh "sudo cat /var/lib/minikube/certs/apiserver.crt" > ~/.minikube/apiserver.crt
> ```
> 
> I tested it from those versions to 1.14.2 with K8s 1.19.3. It works and I believe it will work on Minikube 1.15.0, too.

Thanks a lot, this actually unblocked me.

**Comment by ScrapCodes** [comment]
Above test failure is, due to decommission suite.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. @ScrapCodes .

I also confirmed that the newly added test case passed.

Merged to master for Apache Spark 3.1.0.

**Comment by HyukjinKwon** [comment]
I faced the same problem here https://github.com/apache/spark/pull/30388#issuecomment-728179923 .. It should be best to document this .. It's hard to know from a glance ..

Thanks Dongjoon for workaround here https://github.com/apache/spark/pull/30388#issuecomment-728203026.

**Pull request closed without merging**